src/wattsi.pas

```diff
@@ -495,6 +495,23 @@    TCrossReferences = record
                if (LastSeenHeadingID = '') then
                   LastSeenHeadingID := 'blank-heading';
                LastSeenHeadingID := EnsureID(Element, LastSeenHeadingID).AsString;
+
+               // Append a self-link to each header
+               if (ClassName <> 'no-num no-toc') then
+               begin
+                  HeadingSelfLink := ConstructHTMLElement(eA);
+                  HeadingSelfLink.SetAttribute('class', 'self-link');
+
+                  Scratch := Default(Rope);
+                  Scratch.Append('#');
+                  ExtractedData := Element.GetAttribute('id');
+                  Assert(not ExtractedData.IsEmpty);
+                  Scratch.AppendDestructively(ExtractedData); // appending LastSeenHeadingID does not work
```

Comment:
The reason you can't use LastSeenHeadingID is that it's a local variable, so if you appended it, you would end up having the DOM tree contain a pointer to the stack (rather than the heap). The solution is to not use AsString on the output of EnsureID. Instead, get back a CutRope, which contains pointers to the heap.

Note that EnsureID is itself just going GetAttribute, so it wouldn't make much difference in practice. Also be aware that you can only use AppendDestructively once per reference (hence the name), so if you use LastSeenHeadingID again you'd have to get it back out of the DOM anyway.